### ISSUE TITLE
Different result on same model

### ISSUE THREAD

**Issue by Temicia** [espnet/espnet#2991]
I want same result when running same code. But, sometimes I get different result on same code (I don't change any code) 
In results/log, initial log is same. However, The loss was starting to change different from each other in random iter. and then printed different result

###  first log

        "main/loss_ctc": 35.50965264320374,
        "main/loss_att": 6.756946347951889,
        "main/acc": 0.9789373746701803,
        "main/loss": 15.382758617401123,
        "epoch": 35,
        "iteration": 43300,
        "elapsed_time": 25990.867532494944
    },
    {
        "main/loss_ctc": 33.910652770996094,   << change start
        "main/loss_att": 6.0532202982902525,
        "main/acc": 0.9811846722813258,
        "main/loss": 14.410450391769409,
        "main/cer_ctc": 0.08174386920980926,
        "main/cer": 0.04904632152588556,
        "main/wer": 0.21428571428571427,
        "validation/main/loss_ctc": 47.40722378817472,
        "validation/main/loss_att": 13.899571245366877,
        "validation/main/acc": 0.9586142668241157,
        "validation/main/cer_ctc": 0.1103905811977088,
        "validation/main/cer": 0.05638941539226244,
        "validation/main/wer": 0.19336354368844888,
        "validation/main/loss": 23.951867508165765,
        "epoch": 36,
        "iteration": 43400,
        "elapsed_time": 26296.76114352909
    },
    {
        "main/loss_ctc": 33.4200412940979,
        "main/loss_att": 5.797050931453705,
        "main/acc": 0.9816893570751283,
        "main/loss": 14.083948440551758,
        "epoch": 36,
        "iteration": 43500,
        "elapsed_time": 26348.630531612085
    },

###  second log
        "main/loss_ctc": 35.50965264320374,
        "main/loss_att": 6.756946347951889,
        "main/acc": 0.9789373746701803,
        "main/loss": 15.382758617401123,
        "epoch": 35,
        "iteration": 43300,
        "elapsed_time": 25517.362454279093
    },
    {
        "main/loss_ctc": 34.38905382156372,    << change start
        "main/loss_att": 6.040734493732453,
        "main/acc": 0.9813067018778391,
        "main/loss": 14.545230674743653,
        "main/cer_ctc": 0.08174386920980926,
        "main/cer": 0.04904632152588556,
        "main/wer": 0.21428571428571427,
        "validation/main/loss_ctc": 47.40722378817472,
        "validation/main/loss_att": 13.899571245366877,
        "validation/main/acc": 0.9586142668241157,
        "validation/main/cer_ctc": 0.1103905811977088,
        "validation/main/cer": 0.05638941539226244,
        "validation/main/wer": 0.19336354368844888,
        "validation/main/loss": 23.951867508165765,
        "epoch": 36,
        "iteration": 43400,
        "elapsed_time": 25782.44311495498
    },
    {
        "main/loss_ctc": 33.11112590789795,
        "main/loss_att": 5.5369711530208585,
        "main/acc": 0.983017126241738,
        "main/loss": 13.809217910766602,
        "epoch": 36,
        "iteration": 43500,
        "elapsed_time": 25822.91836479795
    },

Model was trained with WSJ egs (ASR)

and My version are written below

espnet version 0.9.7
pytorch 1.6.0
CUDA 10.1

Thank you.

**Comment by sw005320** [comment]
Hi @Temicia, thanks for the report.
It is actually a known issue.
We were trying to make them at least in the first forward computation or inference, but the backpropagation part cannot be well handled.
That means that unfortunately, we'd have some variance in the final performance, but at least the inference part is deterministic. 
Also, the first forward part is deterministic and you could debug some basic issues there.
(In other words, debugging in the backpropagation part and something in later iterations are not very easy due to indeterministic behaviors... This is a common issue in our current deep learning framework).

**Comment by Temicia** [comment]
Thank you for reply
One computer has different result, but the other has same result result on same script.
Can it be different for each computer?

And sometime CTC loss was negative in pytorch 1.6.0. When I use pytorch 1.0.1, this problem did not occur.

### train.log

    {
        "main/loss_ctc": 86.15477745056153,
        "main/loss_att": 26.781768016815185,
        "main/acc": 0.9224908095354906,
        "main/loss": 44.59367151260376,
        "epoch": 8,
        "iteration": 10100,
        "elapsed_time": 6042.530604951986
    },
    {
        "main/loss_ctc": -1.0633823332454027e+35,
        "main/loss_att": 24.10297875404358,
        "main/acc": 0.9237161502407392,
        "main/loss": -3.190146999736208e+34,
        "epoch": 8,
        "iteration": 10200,
        "elapsed_time": 6084.5271309309755
    },
    {
        "main/loss_ctc": 84.3693364906311,
        "main/loss_att": 26.72009838104248,
        "main/acc": 0.9235564119335804,
        "main/loss": 44.014870433807374,
        "epoch": 8,
        "iteration": 10300,
        "elapsed_time": 6126.193319893995
    },

**Comment by sw005320** [comment]
> Thank you for reply
> One computer has different result, but the other has same result result on same script.
> Can it be different for each computer?

Yes, it happens.

For our record, could you share which GPU card, CUDA version, CuDNN version, and PyTorch version did you use?
(If you follow our bug report template (https://github.com/espnet/espnet/issues/new/choose), you can get such information).

**Comment by Temicia** [comment]
I use

espnet version 0.9.7
pytorch 1.6.0
CUDA 10.1
GPU 1080 Ti
CuDNN 8.0.4

Thank you
